Kubeflow credentials, tarfile fixes, unit tests

rpasricha opened the pull request:
A few combined changes in this PR
* Separate tests into unit and integration (only run Travis for unit tests)
* Authenticate with Kubeflow cluster in Cloud Build to run integration tests
* Add integration test with Kubernetes job deployment
* Mount Kubeflow service account credentials in TFJob (Fixes #106)
* Skip duplicate file additions when generating context tarfile (Fixes #101)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/fairing/112)
<!-- Reviewable:end -->

rpasricha commented:
I'm getting the following test failure for the integration test:

```
Traceback (most recent call last):
  File "/app/function_shim.py", line 15, in <module>
    call(args.serialized_fn...ion_shim.py", line 6, in call
    fn = cloudpickle.load(f)
ImportError: No module named \'test_kubeflow_training\'
```

@r2d4 @karthikv2k Any ideas?

karthikv2k commented:
this is due to the function not in __main__ module. We should set the fn.__module__ = "__main__" in the function preprocessor before serialization. This is not an issue when users are working in the notebooks as all the things defined in a notebook is in __main__ module. This issue mainly occurs while testing.

rpasricha commented:
Sounds good, I'll update the module of the object in the integration test.

rpasricha commented:
Added main module, and fixed another test failure.

The default preprocessor `output_map` argument was an empty dictionary, which is mutable, so file paths from one preprocess call were getting persisted to the next. Updated to use `None` as the default instead of an empty dict.

hamedhsn commented on fairing/deployers/tfjob/tfjob.py:
```diff
@@ -23,6 +23,31 @@ def generate_deployment_spec(self, pod_template_spec):
 
+        # Set appropriate secrets and volumes to enable kubeflow-user service
+        # account.
+        env_var = k8s_client.V1EnvVar(
```
This will break in other platforms than gcp? cant this be optional?
same for volume_mount

r2d4 replied:
I agree, we'll need to add this as something configurable in the deployer

karthikv2k replied:
we should probably have a global config that sets the deployment is happening in GCP and necessary things like gcloud for the builder and auth setup for the deployers are automatically configured.

hamedhsn replied:
global config sets for deployer make sense. then we could also add other configurations like different custom labels for different jobs, volume, gpu setting etc.

r2d4 replied:
Right now tfjob superclasses job so maybe we just have an actual base deployer like we do for builders

rpasricha replied:
Right, thanks for the catch! In order to keep this PR manageable, I'll plan to add a flag to the TFJob deployer, and file a FR for the global config sets.

With the config sets it'll likely be useful to have a base deployer as well, in order for changes like these to be shared accordingly.

rpasricha replied:
Created #118

karthikv2k commented on fairing/preprocessors/base.py:
```diff
@@ -61,9 +61,13 @@ def context_map(self):
     def context_tar_gz(self, output_file=constants.DEFAULT_CONTEXT_FILENAME):
         self.input_files = self.preprocess()
-        with tarfile.open(output_file, "w:gz") as tar:
+        added_files = set()
```
I see it is used in line 68.

r2d4 replied:
ack

rpasricha replied:
Yes, I found a quite a few places where duplicate files can be added to the tarfile, so this is a check to make sure that duplicates are skipped when we are creating the context package itself.

The context map also means that it is possible for multiple input files to map to a single file in the archive, so we store the set of filenames that have actually been added.

r2d4 replied:
Can we move this to be a set in the init function? It's a bit weird to have it be a map, but enforce set qualities later on.

rpasricha replied:
That sounds reasonable, so we can essentially merge `context_map` and `added_files` into a single set. This will also help to keep the source / destination filenames consistent.

rpasricha replied:
Updated to reverse the context map direction to avoid duplicates (destination -> source file), and also add warning messages when duplicate files are skipped.

karthikv2k commented on tests/integration/test_kubeflow_training.py:
```diff
@@ -0,0 +1,32 @@
+    output = sys.stdout.getvalue()
+    sys.stdout = sys.__stdout__
+
+    assert 'train_fn' in output
```
it is better to have printed value other than the function name. I can think of instances where the error message/stack trace have the function name that will give a false test result.

rpasricha replied:
Sure, I'll update the output to use a different string

karthikv2k commented:
/hold

karthikv2k commented:
/hold cancel
/lgtm
/approve
